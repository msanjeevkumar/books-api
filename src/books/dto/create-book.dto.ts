import { ApiProperty } from '@nestjs/swagger';
import * as Joi from 'joi';

export const createBookSchema = Joi.object({
  title: Joi.string().required(),
  author: Joi.string().required(),
  publishedYear: Joi.number()
    .integer()
    .min(1000)
    .max(new Date().getFullYear())
    .required(),
}).strict(true);

export class CreateBookDto {
  @ApiProperty({
    description: 'Title of the book',
    example: 'The Great Gatsby',
  })
  title: string;

  @ApiProperty({
    description: 'Author of the book',
    example: 'F. Scott Fitzgerald',
  })
  author: string;

  @ApiProperty({
    description: 'Year of publication',
    example: 1925,
    maximum: new Date().getFullYear(),
    minimum: 1000,
  })
  publishedYear: number;
}
