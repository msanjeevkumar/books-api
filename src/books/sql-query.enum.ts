export enum SqlQuery {
  InsertBook = 'INSERT INTO books_db.books (id, title, author, publishedYear) VALUES (:id, :title, :author, :publishedYear)',
  SelectAllBooks = 'SELECT * FROM books_db.books',
  SelectBookById = 'SELECT * FROM books_db.books WHERE id = :id',
  UpdateBook = 'UPDATE books_db.books SET {fields} WHERE id = :id',
  DeleteBook = 'DELETE FROM books_db.books WHERE id = :id',
}
