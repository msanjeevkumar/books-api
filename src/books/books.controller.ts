import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  HttpCode,
  UsePipes,
} from '@nestjs/common';
import { BooksService } from './books.service';
import { CreateBookDto, createBookSchema } from './dto/create-book.dto';
import { UpdateBookDto, updateBookDtoSchema } from './dto/update-book.dto';
import {
  ApiBadRequestResponse,
  ApiCreatedResponse,
  ApiNoContentResponse,
  ApiNotFoundResponse,
  ApiOkResponse,
  ApiParam,
  ApiTags,
} from '@nestjs/swagger';
import { Book } from './entities/book.entity';
import { JoiValidationPipe } from '../joi.validation.pipe';

@Controller('api/books')
@ApiTags('books')
export class BooksController {
  constructor(private readonly booksService: BooksService) {}

  @Post()
  @ApiCreatedResponse({ description: 'Book created successfully' })
  @ApiBadRequestResponse({
    description: 'Validation failed',
    schema: {
      example: {
        message: '"title" is required',
        error: 'Bad Request',
        statusCode: 400,
      },
    },
  })
  async create(
    @Body(new JoiValidationPipe(createBookSchema)) createBookDto: CreateBookDto,
  ) {
    return this.booksService.create(createBookDto);
  }

  @Get()
  @ApiOkResponse({ type: [Book] })
  async findAll() {
    return this.booksService.findAll();
  }

  @Get(':id')
  @ApiOkResponse({ type: Book })
  @ApiNotFoundResponse({
    description: 'Book not found',
    schema: {
      example: {
        statusCode: 404,
        message: 'Book with id:d170d408-6330-448b-9461-a55696fe4d69 not found',
        error: 'Not Found',
      },
    },
  })
  @ApiParam({
    name: 'id',
    type: String,
    description: 'unique identifier of the book',
    example: 'd170d408-6330-448b-9461-a55696fe4d69',
  })
  async findOne(@Param('id') id: string) {
    return this.booksService.findOne(id);
  }

  @Patch(':id')
  @HttpCode(204)
  @ApiBadRequestResponse({
    description: 'Validation failed',
    schema: {
      example: {
        statusCode: 400,
        message:
          '"value" must contain at least one of [title, author, publishedYear]',
        error: 'Bad Request',
      },
    },
  })
  @ApiNoContentResponse({ description: 'Book updated successfully' })
  @ApiNotFoundResponse({
    description: 'Book not found',
    schema: {
      example: {
        statusCode: 404,
        message: 'Book with id:d170d408-6330-448b-9461-a55696fe4d69 not found',
        error: 'Not Found',
      },
    },
  })
  async update(
    @Param('id') id: string,
    @Body(new JoiValidationPipe(updateBookDtoSchema))
    updateBookDto: UpdateBookDto,
  ) {
    return this.booksService.update(id, updateBookDto);
  }

  @Delete(':id')
  @HttpCode(204)
  @ApiNoContentResponse({ description: 'Book deleted successfully' })
  @ApiNotFoundResponse({
    description: 'Book not found',
    schema: {
      example: {
        statusCode: 404,
        message: 'Book with id:d170d408-6330-448b-9461-a55696fe4d69 not found',
        error: 'Not Found',
      },
    },
  })
  async remove(@Param('id') id: string) {
    return this.booksService.remove(id);
  }
}
