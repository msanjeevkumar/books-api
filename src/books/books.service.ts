import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { uuid } from 'uuidv4';
import { CreateBookDto } from './dto/create-book.dto';
import { UpdateBookDto } from './dto/update-book.dto';
import { Sequelize } from 'sequelize-typescript';
import { QueryTypes } from 'sequelize';
import { Book } from './entities/book.entity';

@Injectable()
export class BooksService {
  constructor(private readonly sequelize: Sequelize) {}

  async create(createBookDto: CreateBookDto) {
    const insertQuery = `
        INSERT INTO books_db.books (id, title, author, publishedYear)
        VALUES (:id, :title, :author, :publishedYear);
    `;

    const book = { id: uuid(), ...createBookDto };
    await this.sequelize.query(insertQuery, {
      type: QueryTypes.INSERT,
      replacements: {
        ...book,
      },
    });

    return book;
  }

  async findAll(): Promise<Book[]> {
    const selectAllQuery = `SELECT * FROM books_db.books`;
    return this.sequelize.query(selectAllQuery, {
      type: QueryTypes.SELECT,
    });
  }

  async findOne(id: string): Promise<Book> {
    const selectOneQuery = `SELECT * FROM books_db.books WHERE id = :id`;
    const book: Book = await this.sequelize.query(selectOneQuery, {
      type: QueryTypes.SELECT,
      replacements: { id },
      plain: true,
    });

    if (!book) {
      this.throwNotFoundException(id);
    }

    return book;
  }

  async update(id: string, updateBookDto: UpdateBookDto): Promise<void> {
    if (Object.keys(updateBookDto).length === 0) {
      throw new BadRequestException('At least one field is needed to update');
    }

    const existingBook = await this.findOne(id);

    const fieldMappings = [
      { key: 'title', column: 'title' },
      { key: 'author', column: 'author' },
      {
        key: 'publishedYear',
        column: 'publishedYear',
      },
    ];

    const updateFields = fieldMappings
      .filter(
        ({ key }) =>
          updateBookDto[key] !== undefined &&
          updateBookDto[key] !== existingBook[key],
      )
      .map((mapping) => `${mapping.column} = :${mapping.key}`);

    if (updateFields.length === 0) {
      return; // No updates needed
    }

    const updateQuery = `
      UPDATE books_db.books
      SET
        ${updateFields.join(', ')}
      WHERE id = :id
    `;

    const replacements = { id, ...updateBookDto };

    await this.sequelize.query(updateQuery, {
      type: QueryTypes.UPDATE,
      replacements,
    });
  }

  async remove(id: string): Promise<void> {
    const deleteQuery = `DELETE FROM books_db.books WHERE id = :id`;
    const [, meta] = (await this.sequelize.query(deleteQuery, {
      type: QueryTypes.RAW,
      replacements: { id },
    })) as [any, { affectedRows: number }];

    if (meta.affectedRows === 0) {
      this.throwNotFoundException(id);
    }
  }

  private throwNotFoundException(id: string) {
    throw new NotFoundException(`Book with id:${id} not found`);
  }
}
