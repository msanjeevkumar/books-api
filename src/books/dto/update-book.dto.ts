import { PartialType } from '@nestjs/swagger';
import { CreateBookDto } from './create-book.dto';
import * as Joi from 'joi';

export const updateBookDtoSchema = Joi.object({
  title: Joi.string().optional(),
  author: Joi.string().optional(),
  publishedYear: Joi.number()
    .integer()
    .min(1000)
    .max(new Date().getFullYear())
    .optional(),
}).or('title', 'author', 'publishedYear');

export class UpdateBookDto extends PartialType(CreateBookDto) {}
