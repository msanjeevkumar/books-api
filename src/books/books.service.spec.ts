import { Test, TestingModule } from '@nestjs/testing';
import { BooksService } from './books.service';
import { Sequelize } from 'sequelize-typescript';
import { CreateBookDto } from './dto/create-book.dto';
import { QueryTypes } from 'sequelize';
import { uuid } from 'uuidv4';
import { Book } from './entities/book.entity';
import { NotFoundException } from '@nestjs/common';
import { UpdateBookDto } from './dto/update-book.dto';
import { SqlQuery } from './sql-query.enum';
jest.mock('uuidv4');

describe('BooksService', () => {
  let booksService: BooksService;
  let sequelizeMock: any;

  beforeEach(async () => {
    sequelizeMock = {
      query: jest.fn(),
    };

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        BooksService,
        { provide: Sequelize, useValue: sequelizeMock },
      ],
    }).compile();

    booksService = module.get<BooksService>(BooksService);
  });

  describe('create', () => {
    const createBookDto: CreateBookDto = {
      title: 'Test Book',
      author: 'Test Author',
      publishedYear: 2023,
    };
    const mockUuid = 'generated-id';

    beforeEach(() => {
      (uuid as jest.Mock).mockReturnValue(mockUuid);
    });

    it('should create a book and return it', async () => {
      // Arrange
      sequelizeMock.query.mockResolvedValueOnce([]);

      // Act
      const result = await booksService.create(createBookDto);

      // Assert
      expect(result).toMatchSnapshot(); // Use inline snapshot
      expect(sequelizeMock.query).toHaveBeenCalledWith(
        SqlQuery.InsertBook,
        expect.objectContaining({
          type: QueryTypes.INSERT,
          replacements: {
            id: mockUuid,
            ...createBookDto,
          },
        }),
      );
    });

    it('should handle errors during creation', async () => {
      // Arrange
      const errorMessage = 'Failed to insert';
      sequelizeMock.query.mockRejectedValueOnce(new Error(errorMessage));

      // Act & Assert
      await expect(booksService.create(createBookDto)).rejects.toThrowError(
        errorMessage,
      );
      expect(sequelizeMock.query).toHaveBeenCalledWith(
        SqlQuery.InsertBook,
        expect.objectContaining({
          type: QueryTypes.INSERT,
          replacements: {
            ...createBookDto,
            id: mockUuid,
          },
        }),
      );
    });
  });

  describe('findAll', () => {
    it('should return an array of books', async () => {
      // Arrange
      const expectedBooks: Book[] = [
        {
          id: 'ede9fa21-9fd0-4e7d-8e80-a39f3dcdbe7c',
          title: 'Book 1',
          author: 'Author 1',
          publishedYear: 2020,
        },
        {
          id: 'e6ad7085-4967-4e11-be20-e4d1e6c5d716',
          title: 'Book 2',
          author: 'Author 2',
          publishedYear: 2021,
        },
      ];
      sequelizeMock.query.mockResolvedValue(expectedBooks);

      // Act
      const result = await booksService.findAll();

      // Assert
      expect(result).toEqual(expectedBooks);
      expect(sequelizeMock.query).toHaveBeenCalledWith(
        SqlQuery.SelectAllBooks,
        expect.objectContaining({
          type: QueryTypes.SELECT,
        }),
      );
    });

    it('should handle empty result and return an empty array', async () => {
      // Arrange
      sequelizeMock.query.mockResolvedValue([]);

      // Act
      const result = await booksService.findAll();

      // Assert
      expect(result).toEqual([]);
      expect(sequelizeMock.query).toHaveBeenCalledWith(
        SqlQuery.SelectAllBooks,
        expect.objectContaining({
          type: QueryTypes.SELECT,
        }),
      );
    });
  });

  describe('findOne', () => {
    it('should find and return a book by id', async () => {
      // Arrange
      const bookId = 'valid-book-id';
      const expectedBook: Book = {
        id: bookId,
        title: 'Sample Book',
        author: 'John Doe',
        publishedYear: 2023,
      };
      sequelizeMock.query.mockResolvedValue(expectedBook);

      // Act
      const result = await booksService.findOne(bookId);

      // Assert
      expect(result).toEqual(expectedBook);
      expect(sequelizeMock.query).toHaveBeenCalledWith(
        SqlQuery.SelectBookById,
        {
          type: QueryTypes.SELECT,
          replacements: { id: bookId },
          plain: true,
        },
      );
    });

    it('should throw NotFoundException if book is not found', async () => {
      // Arrange
      const bookId = 'non-existent-book-id';
      sequelizeMock.query.mockResolvedValue(null);

      // Act and Assert
      await expect(booksService.findOne(bookId)).rejects.toThrowError(
        NotFoundException,
      );

      expect(sequelizeMock.query).toHaveBeenCalledWith(
        SqlQuery.SelectBookById,
        {
          type: QueryTypes.SELECT,
          replacements: { id: bookId },
          plain: true,
        },
      );
    });
  });

  describe('update', () => {
    it('should not update if the new values are the same as existing values of the book', async () => {
      // Arrange
      const id = 'book-id';
      const updateBookDto: UpdateBookDto = {
        title: 'Old Title',
        author: 'Author',
        publishedYear: 2020,
      };
      const existingBook: Book = {
        id,
        title: 'Old Title',
        author: 'Author',
        publishedYear: 2020,
      };
      sequelizeMock.query.mockResolvedValueOnce(existingBook);

      // Act
      const result = await booksService.update(id, updateBookDto);

      // Assert
      expect(result).toBeUndefined();
      expect(sequelizeMock.query).toHaveBeenCalledTimes(1);
      expect(sequelizeMock.query).toHaveBeenCalledWith(
        SqlQuery.SelectBookById,
        {
          type: QueryTypes.SELECT,
          replacements: { id },
          plain: true,
        },
      );
    });

    it('should throw NotFoundException if the book is not found', async () => {
      // Arrange
      const id = 'non-existent-id';
      const updateBookDto: UpdateBookDto = { title: 'New Title' };
      sequelizeMock.query.mockResolvedValue(null);

      // Act and Assert
      await expect(booksService.update(id, updateBookDto)).rejects.toThrowError(
        NotFoundException,
      );

      expect(sequelizeMock.query).toHaveBeenCalledTimes(1);
      expect(sequelizeMock.query).toHaveBeenCalledWith(
        SqlQuery.SelectBookById,
        {
          type: QueryTypes.SELECT,
          replacements: { id },
          plain: true,
        },
      );
    });

    it('should update the specified fields', async () => {
      // Arrange
      const updateBookDto: UpdateBookDto = {
        title: 'Updated Title',
        author: 'Updated Author',
      };
      const existingBook: Book = {
        id: 'existing-id',
        title: 'Existing Book',
        author: 'Existing Author',
        publishedYear: 2020,
      };
      const expectedQuery = `UPDATE books_db.books SET title = :title, author = :author WHERE id = :id`;
      const expectedReplacements = {
        id: existingBook.id,
        ...updateBookDto,
      };
      sequelizeMock.query.mockResolvedValueOnce(existingBook);
      sequelizeMock.query.mockResolvedValueOnce({ affectedRows: 1 });

      // Act
      await booksService.update(existingBook.id, updateBookDto);

      // Assert
      expect(sequelizeMock.query).toHaveBeenNthCalledWith(
        1,
        SqlQuery.SelectBookById,
        {
          type: QueryTypes.SELECT,
          replacements: { id: existingBook.id },
          plain: true,
        },
      );
      expect(sequelizeMock.query).toHaveBeenNthCalledWith(2, expectedQuery, {
        type: QueryTypes.UPDATE,
        replacements: expectedReplacements,
      });
    });
  });

  describe('remove', () => {
    it('should remove a book', async () => {
      // Arrange
      const id = 'test-id';
      const affectedRows = 1;
      const mockQueryResult = [[], { affectedRows }];
      sequelizeMock.query.mockResolvedValueOnce(mockQueryResult);

      // Act
      await booksService.remove(id);

      // Assert
      expect(sequelizeMock.query).toHaveBeenCalledWith(SqlQuery.DeleteBook, {
        type: QueryTypes.RAW,
        replacements: { id },
      });
    });

    it('should throw NotFoundException if book is not found', async () => {
      // Arrange
      const id = 'test-id';
      const affectedRows = 0;
      const mockQueryResult = [[], { affectedRows }];
      sequelizeMock.query.mockResolvedValueOnce(mockQueryResult);

      // Act and Assert
      await expect(booksService.remove(id)).rejects.toThrowError(
        NotFoundException,
      );
      expect(sequelizeMock.query).toHaveBeenCalledWith(SqlQuery.DeleteBook, {
        type: QueryTypes.RAW,
        replacements: { id },
      });
    });

    it('should handle database error', async () => {
      // Arrange
      const id = 'test-id';
      const errorMessage = 'Database error';
      sequelizeMock.query.mockRejectedValueOnce(new Error(errorMessage));

      // Act and Assert
      await expect(booksService.remove(id)).rejects.toThrowError(errorMessage);
      expect(sequelizeMock.query).toHaveBeenCalledWith(SqlQuery.DeleteBook, {
        type: QueryTypes.RAW,
        replacements: { id },
      });
    });
  });
});
