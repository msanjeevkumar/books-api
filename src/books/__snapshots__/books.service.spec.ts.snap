// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksService create should create a book and return it 1`] = `
{
  "author": "Test Author",
  "id": "generated-id",
  "publishedYear": 2023,
  "title": "Test Book",
}
`;
